Base Implementation of schema.org

shazmasiddiqui opened the pull request:
## Introduction

To ensure the delivery of high-quality search results, web crawlers rely on structured data that is used by search engines to generate rich snippets. Currently, Joomla! provides a feature for generating rich snippets, but it uses inline microdata that is challenging to modify from the backend since it is hardcoded into the HTML.

However, a proposed solution is to introduce schema.org and JSON-LD implementation to Joomla!. This would allow structured data to be added and configured via Joomla! backend, simplifying the optimization of website content for search engines and improving the visibility of the website in search results.

By incorporating this update, Joomla! users would benefit from an easier and more flexible way to optimize their website's structured data, leading to a better search experience for their users and potentially increasing their website's traffic and engagement.

## Summary of Changes

1. **Schemaorg System Plugin:** This plugin creates a base form for adding schema. It will generate a default form with an empty list to select a schema type, and then generate a form with relevant attributes for that schema type. This plugin will provide a more user-friendly way of adding structured data to Joomla websites, without requiring users to manually code HTML. The plugin will also use JSON-LD markup for schema implementation, making it easier to read and maintain than traditional HTML markup.

2. **Schemaorg Plugins:** Each schema type is implemented as a separate plugin. It allows for a more modular and extensible architecture. Each plugin is responsible for generating the schema markup for a specific type of content, which allows for more flexibility and customization options for website owners. It makes it easier for third-party developers to contribute to the development of Joomla's structured data capabilities.
Plugins that we currently have are:
- Blogposting Plugin
- Book Plugin
- Event Plugin
- Organization Plugin
- Person Plugin
- Recipe Plugin
- JobPosting Plugin (Thanks to @tfeuerst)

3. **Database:**
- Adds a `#__schemaorg` database table.
- Adds `#__extensions` entries for `plg_system_schemaorg`, `plg_schemaorg_blogposting`, `plg_schemaorg_book`, `plg_schemaorg_event`, `plg_schemaorg_organization`, `plg_schemaorg_person`, `plg_schemaorg_recipe`, `plg_schemaorg_jobposting`.

## Future Ideas

- [ ] Implement an automation feature to pre-fill common data automatically in the Schemaorg form.
- [ ] Enhance the backend interface to allow users to review and customize pre-filled data as needed.
- [ ] Implement a Schema Testing feature within the Schemaorg System Plugin to validate the generated schema.
- [ ] Conduct thorough testing, bug fixing, and integration testing.
- [ ] Prepare documentation. 

## Testing Instructions
For testing, a fresh installation/config is important. Steps:
1. Go to System Dashboard > Manage module > Plugins
<img width="1256" alt="Screenshot 2023-07-11 at 11 42 37 PM" src="https://github.com/joomla-projects/gsoc22_schema.org/assets/43249198/6a55beab-fdf2-4c71-b04e-7d3cea266ed7">

2. Search for "Schema.org System Plugin"
<img width="1256" alt="Screenshot 2023-07-11 at 11 44 59 PM" src="https://github.com/joomla-projects/gsoc22_schema.org/assets/43249198/d6873a4c-d875-4cf5-a325-1be03764df31">

3. Configure the plugin by adding representation schema
<img width="1256" alt="Screenshot 2023-07-11 at 11 38 19 AM" src="https://github.com/joomla-projects/gsoc22_schema.org/assets/43249198/65cb3157-cd99-4b85-bd2e-46923f398632">

4. Go to Content > Articles and click on “New” to add a new article
<img width="1256" alt="Screenshot 2023-07-11 at 11 14 43 AM" src="https://github.com/joomla-projects/gsoc22_schema.org/assets/43249198/aad1ccad-da9c-4bd8-b872-20279978c786">

5. Select Schema type
<img width="1256" alt="Screenshot 2023-07-11 at 11 58 40 AM" src="https://github.com/joomla-projects/gsoc22_schema.org/assets/43249198/e4134f64-99df-40c7-8ecb-8d8bf67236a5">

6. Fill in all the details
![Screenshot_scroll](https://github.com/joomla-projects/gsoc22_schema.org/assets/43249198/721f6396-a66d-4491-9830-8f91f113f346)

7. Schema is generated and added to the head tag of the html document in the frontend
<img width="1256" alt="Screenshot 2023-07-11 at 1 01 27 PM" src="https://github.com/joomla-projects/gsoc22_schema.org/assets/43249198/fcbb7713-c218-466a-9879-71603bcf8cbc">

<img width="1023" alt="Screenshot 2023-07-11 at 1 02 25 PM" src="https://github.com/joomla-projects/gsoc22_schema.org/assets/43249198/6459a7dc-f009-4715-9631-32f0c6e3b966">

## Testing Generated Schema using Google's structured data testing tool

**Generated Schema**
```
{
  "@context": "https://schema.org",
  "@graph": [
    {
      "@type": "Organization",
      "@id": "http://localhost/gsoc22_schema.org/#/schema/Organization/base",
      "name": "Joomla Test",
      "url": "http://localhost/gsoc22_schema.org/",
      "logo": {
        "@type": "ImageObject",
        "@id": "http://localhost/gsoc22_schema.org/#/schema/ImageObject/logo",
        "url": "images/joomla_black.png",
        "contentUrl": "images/joomla_black.png",
        "width": 225,
        "height": 50
      },
      "image": {
        "@id": "http://localhost/gsoc22_schema.org/#/schema/ImageObject/logo"
      }
    },
    {
      "@type": "WebSite",
      "@id": "http://localhost/gsoc22_schema.org/#/schema/WebSite/base",
      "url": "http://localhost/gsoc22_schema.org/",
      "name": "Schemaorg Demo",
      "publisher": {
        "@id": "http://localhost/gsoc22_schema.org/#/schema/Organization/base"
      }
    },
    {
      "@type": "WebPage",
      "@id": "http://localhost/gsoc22_schema.org/#/schema/WebPage/base",
      "url": "http://localhost/gsoc22_schema.org/index.php/articles/moms-world-famous-banana-bread",
      "name": "Mom's World Famous Banana Bread",
      "description": null,
      "isPartOf": {
        "@id": "http://localhost/gsoc22_schema.org/#/schema/WebSite/base"
      },
      "about": {
        "@id": "http://localhost/gsoc22_schema.org/#/schema/Organization/base"
      },
      "inLanguage": "en-GB",
      "breadcrumb": {
        "@id": "http://localhost/gsoc22_schema.org/#/schema/BreadcrumbList/17"
      }
    },
    {
      "@type": "Recipe",
      "image": "images/Moms%20World%20Famous%20Banana%20Bread.jpeg",
      "name": "Mom's World Famous Banana Bread",
      "author": "John Smith",
      "cookTime": "PT1H",
      "datePublished": "2023-07-11",
      "description": "This classic banana bread recipe comes from my mom -- the walnuts add a nice texture and flavor to the banana bread.",
      "prepTime": "PT15M",
      "recipeInstructions": "Preheat the oven to 350 degrees. Mix in the ingredients in a bowl. Add the flour last. Pour the mixture into a loaf pan and bake for one hour.",
      "recipeYield": "1 loaf",
      "recipeIngredient": [
        "3 or 4 ripe bananas, smashed",
        "1 egg",
        "3/4 cup of sugar"
      ],
      "nutrition": {
        "@type": "NutritionInformation",
        "calories": "240 calories",
        "fatContent": "9 grams fat"
      },
      "@id": "http://localhost/gsoc22_schema.org/#/schema/Recipe/1",
      "@isPartOf": {
        "@id": "http://localhost/gsoc22_schema.org/#/schema/WebPage/base"
      }
    }
  ]
}
```

**Test Results**

<img width="1256" alt="Screenshot 2023-07-11 at 1 21 17 PM" src="https://github.com/joomla-projects/gsoc22_schema.org/assets/43249198/16be7275-b8e1-4f27-b8f1-f123611b1f0e">



## Documentation changes required

- Update the user documentation to include detailed instructions on utilizing the new Schemaorg Plugins.
- Document the process of adding new schema types in the future and how to integrate them with the existing schema implementation.

## Acknowledgements
I would like to express my heartfelt gratitude to my mentors @bembelimen, @ditsuke, @anuragteapot, and @rjharishabh for their continuous guidance and availability throughout the project. I also want to thank @wilsonge for his PR, which served as a valuable reference. Lastly, I am grateful to the Joomla! community for their support and contributions.

shazmasiddiqui committed:
Base Implementation of schema.org

shazmasiddiqui committed:
Adds language strings

shazmasiddiqui committed:
Fixed absolute path for schemaorg.xml

shazmasiddiqui committed:
Bloposting plugin

shazmasiddiqui committed:
Book plugin

shazmasiddiqui committed:
Event Plugin

shazmasiddiqui committed:
Updates language files

shazmasiddiqui committed:
Adds organization schema

bembelimen committed:
Merge pull request #18 from joomla-projects/organisation

Organization plugin

bembelimen committed:
Fix file line end

bembelimen committed:
Add empty line at end of file

bembelimen committed:
Add empty line at end of file

shazmasiddiqui committed:
Merge branch 'base' into blogposting

shazmasiddiqui committed:
Merge branch 'base' into book

bembelimen committed:
Merge pull request #15 from joomla-projects/blogposting

Blogposting Plugin

shazmasiddiqui committed:
Update provider.php

Added empty lines

shazmasiddiqui committed:
Update Book.php

Added empty lines

shazmasiddiqui committed:
Merge branch 'base' into book

shazmasiddiqui committed:
Added empty line at the end of file

bembelimen committed:
Merge pull request #16 from joomla-projects/book

Book plugin

shazmasiddiqui committed:
Resolving conflicts

shazmasiddiqui committed:
Update 4.2.3-2022-12-10.sql

shazmasiddiqui committed:
Update 4.2.3-2022-12-10.sql

shazmasiddiqui committed:
Update base.sql

shazmasiddiqui committed:
Update base.sql

shazmasiddiqui committed:
Resolve conflicts

shazmasiddiqui committed:
Resolving conflicts

shazmasiddiqui committed:
Update 4.2.3-2022-12-10.sql

shazmasiddiqui committed:
Update 4.2.3-2022-12-10.sql

shazmasiddiqui committed:
Update provider.php

shazmasiddiqui committed:
Update Event.php

bembelimen committed:
Merge pull request #17 from joomla-projects/event

Event Plugin

shazmasiddiqui committed:
Adds Person plugin

shazmasiddiqui committed:
Adds empty line

shazmasiddiqui committed:
Adds empty line

shazmasiddiqui committed:
Adds empty line at the end of file

shazmasiddiqui committed:
Update provider.php

bembelimen committed:
Merge pull request #19 from joomla-projects/person

Person Plugin

shazmasiddiqui committed:
Merge branch 'base' of github.com:joomla-projects/gsoc22_schema.org into base

:wq
:wq

bembelimen committed:
Moved xml outside of src folder and load via trait

bembelimen committed:
Remove $supportFunctionality

bembelimen committed:
Move onSchemaPrepareData to trait

bembelimen committed:
Move onSchemaPrepareData to trait

bembelimen committed:
Moved getSubscribedEvents to trait

bembelimen committed:
Move onSchemaAfterSave to trait

shazmasiddiqui committed:
Adds recipe plugin

bembelimen committed:
Use EventInterface as type hint

shazmasiddiqui committed:
Merge pull request #20 from joomla-projects/bembelimen/improvements

Moved xml outside of src folder and load via trait

shazmasiddiqui committed:
Merge branch 'base' of github.com:joomla-projects/gsoc22_schema.org into recipe

bembelimen committed:
Move XML outside of src folder for schema.org system plugin

shazmasiddiqui committed:
Updated file structure

shazmasiddiqui committed:
Update plg_schemaorg_recipe.ini

shazmasiddiqui committed:
Update duration.xml

shazmasiddiqui committed:
Update provider.php

bembelimen committed:
Merge pull request #21 from joomla-projects/recipe

Recipe Plugin

shazmasiddiqui committed:
cleanups

shazmasiddiqui committed:
Adds schemaorg in contacts component

bembelimen commented on administrator/components/com_contact/src/Extension/ContactComponent.php:
```diff
@@ -43,18 +45,26 @@ class ContactComponent extends MVCComponent implements
+        'core.featured' => true,
+        'core.state' => true,
+    ];
+
```
I think you don't need this anymore

shazmasiddiqui replied:
done

bembelimen committed:
Add empty line

richard67 committed:
SQL fixes

- Consistent SQL for new installations and updates
- Larger data type for id and itemId columns
- No display width for integer columns on MySQL
- Use order number for schemaorg plugins
- Fix order numbers of system plugins on new installation
- Fix character set and collation for MySQL on new installation
- Align code style to existing SQL

richard67 committed:
One more CS fix

richard67 committed:
Alphabetically sort new schemaorg plugins

shazmasiddiqui committed:
Update ContactComponent.php

shazmasiddiqui committed:
version updated

bembelimen committed:
Fix language files

shazmasiddiqui committed:
Merge pull request #23 from joomla-projects/bembelimen/languagefiles

Fix language files

shazmasiddiqui committed:
Merge pull request #22 from richard67/base-fix-sql-scripts-by-richard67

SQL fixes for PR #14

bembelimen committed:
Merge branch 'dev' into base

shazmasiddiqui committed:
Deleting rich snippets from components/com_content folder

shazmasiddiqui committed:
Removed rich snippets from contact component

shazmasiddiqui committed:
Removes rich snippets from layout folder

shazmasiddiqui committed:
minor fixes

shazmasiddiqui committed:
Resolved merge conflicts

shazmasiddiqui committed:
Fixed phpcs

shazmasiddiqui committed:
Resolved conflicts

bembelimen committed:
Merge branch 'dev' into base

bembelimen committed:
Fix code style

bembelimen committed:
Improve information for schema and require base configuration

bembelimen committed:
Fix basetype parameter

bembelimen committed:
Added base type in output

bembelimen committed:
Remove debug output

bembelimen committed:
Add breadcrumb and finder implementation

bembelimen committed:
Add image and social media

bembelimen committed:
Check for context if schema is supported for the extension

bembelimen committed:
CS Fix

bembelimen committed:
CS fixes

bembelimen committed:
CS fix

bembelimen committed:
Fix deploy version

bembelimen committed:
Improve data loading

bembelimen committed:
Improve trait

bembelimen committed:
Fix broken variables

bembelimen committed:
CS fix

bembelimen committed:
Merge remote-tracking branch 'origin/base' into bembelimen/baseinformation

bembelimen committed:
CS fix

bembelimen committed:
CS fix

bembelimen committed:
CS fix

bembelimen committed:
Fix from type

bembelimen committed:
Cleanup plugins

bembelimen committed:
Add base structure of blogposting

bembelimen committed:
Add JobPosting installer

bembelimen committed:
Fix CS

bembelimen committed:
Fix new line

Commit:
Form fields for jobposting

bembelimen committed:
Merge branch 'bembelimen/baseinformation' into tfeuerst/jobposting

bembelimen committed:
Fix structure

shazmasiddiqui committed:
Merge pull request #25 from joomla-projects/bembelimen/baseinformation

Implement base information organization/person

Commit:
Provide form for schema.org JoPosting

Commit:
Fix linebreaks

bembelimen committed:
Remove unused namespace

shazmasiddiqui committed:
Changed schema structure and fixed the customCleanup

shazmasiddiqui committed:
Added generic field

shazmasiddiqui committed:
Merge pull request #26 from joomla-projects/tfeuerst/jobposting

Add jobposting fields

shazmasiddiqui committed:
Changed input type for Nutrition fields to text
